web/packages/teleport/src/services/api/api.js

```diff
@@ -64,52 +65,56 @@ const api = {
     });
   },
 
-  fetchJson(url, params) {
-    return new Promise((resolve, reject) => {
-      this.fetch(url, params)
-        .then(response => {
-          if (response.ok) {
-            return response
-              .json()
-              .then(json => resolve(json))
-              .catch(err =>
-                reject(new ApiError(err.message, response, { cause: err }))
-              );
-          } else {
-            return response
-              .json()
-              .then(json => reject(new ApiError(parseError(json), response)))
-              .catch(err => {
-                reject(
-                  new ApiError(
-                    `${response.status} - ${response.url}`,
-                    response,
-                    { cause: err }
-                  )
-                );
-              });
-          }
-        })
-        .catch(err => {
-          reject(err);
-        });
-    });
+  async fetchJson(url, params) {
+    let response;
+    try {
+      response = await this.fetch(url, params);
+      let json = await response.json();
+
+      // Retry with MFA if we get an admin action missing MFA error.
+      if (!response.ok && isAdminActionRequiresMFAError(parseError(json))) {
+        params.headers = {
+          ...params.headers,
+          'Mfa-Response': JSON.stringify({
+            webauthnAssertionResponse: await auth.getWebauthnResponse(),
+          }),
+        };
+        response = await this.fetch(url, params);
+        json = await response.json();
+      }
+
+      if (!response.ok) {
+        throw new ApiError(parseError(json), response);
+      }
+
+      return json;
+    } catch (err) {
+      throw new ApiError(err.message, response, { cause: err });
+    }
   },
 
-  fetch(url, params = {}) {
+  async fetch(url, params = {}) {
     url = window.location.origin + url;
     const options = {
       ...requestOptions,
       ...params,
     };
 
     options.headers = {
+      ...requestOptions.headers,
```

Comment:
I don't think we need it? We spread `requestOptions` in line 99 so we should have all the headers in line 105.

Reply:
On line 99 the `requestOptions.headers` were being overwritten by `params.headers` when set in params.